_posts/2018-12-04-imposing-invariants-in-scala.md

```diff
@@ -0,0 +1,359 @@
+    val guestTicket: Seat = reservation.guestTicket.seat
+
+    (userSeat, guestTicket) match {
+      case (_: Standing.type, _: Standing.type ) => "Go to gate 1"
+      case (userSeat: VipTribune, _: VipTribune ) => s"Go to gate 2. Your tribune is named ${userSeat.tribuneId}"
+      case (userSeat: NormalSeat, guestSeat: NormalSeat ) => s"Go to gate 3. You and your guest are in a seats ${userSeat.column} and ${guestSeat.column} respectively"
+    }
+  }
+``` 
+ 
+The code above will work well and do what we are looking for. But we are not quite there yet. The compiler still finds a warning:  
+
+```
+match may not be exhaustive.
+[warn] It would fail on the following inputs: (NormalSeat(_, _), Standing), (NormalSeat(_, _), VipTribune(_)), (Standing, NormalSeat(_, _)), (Standing, VipTribune(_)), (VipTribune(_), NormalSeat(_, _)), (VipTribune(_), Standing)
+[warn]     (userSeat, guestTicket) match {
+[warn]     ^
+```
+
+The compiler is concerned we might not be considering all possible combinations of seats within a reservation.   
+We are not considering for the cross cases (e.g. (Standing, NormalSeat)) because with the above tricks, we are certain that any created instance of Reservation will always respect the invariant. 
+The compiler isn’t. For it, a **single** Reservation[Seat] might contain tickets in different seating areas. We have to match on both tickets (user and guest) each of which, as far as the compiler goes, can be of any type (<code>Standing</code>, <code>VipTribune</code>, <code>NormalSeated</code>, <code>Seat</code>).  
+Because we (arguably) know more than the compiler in this case, we can tell him not to be concerned with the non-exhaustive pattern match with an annotation (chapter 15.5 at <a id="mmds-ref" href="#programming-in-scala">[1]</a>):   
+
+```scala
+((userSeat, guestTicket): @unchecked) match {
+```
+
+Which will prevent the warning during compilation.
+
+
+# 4. Using OOP - Subtyping
+
+This is the approach most of us would follow. Probably the most pragmatic.
+
+It requires having a bare trait Reservation and extending it for every sub-type of Seat. When there are many sub-types, this would become very repetitive.  
+
+```scala
+sealed trait Reservation[+SeatType <: Seat]
+
+case class ReservationSeated(
+  userTicket: Ticket[NormalSeat],
+  guestTicket: Ticket[NormalSeat]
+) extends Reservation[NormalSeat]
+
+case class ReservationStanding(
+  userTicket: Ticket[Standing.type],
+  guestTicket: Ticket[Standing.type ]
+) extends Reservation[Standing.type]
+
+case class ReservationVipTribune(
+  userTicket: Ticket[VipTribune],
+  guestTicket: Ticket[VipTribune]
+) extends Reservation[VipTribune]
+```
+
+On the other hand, all the type information is carried on at run-time. We can for example pattern match with confidence:  
+
+```scala
+def guideUsersToSeat(reservation: Reservation[Seat]): String = reservation match {
+  case ReservationStanding(_, _) => s"Go to gate 1."
```

Comment:
No need for String interpolation here. I only raised this because you don't have that here:
> case _: Reservation[Standing.type] => "Go to gate 1. You and your guest must stand."

and 
> case (_: Standing.type, _: Standing.type ) => "Go to gate 1"

Reply:
Good eye.